packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.kt

```diff
@@ -151,23 +153,32 @@ class DeviceModule : Module() {
         ?: return DeviceType.UNKNOWN
 
       // Get display metrics to see if we can differentiate phones and tablets.
-      val metrics = DisplayMetrics()
-      windowManager.defaultDisplay.getMetrics(metrics)
-
-      // Calculate physical size.
-      val widthInches = metrics.widthPixels / metrics.xdpi.toDouble()
-      val heightInches = metrics.heightPixels / metrics.ydpi.toDouble()
-      val diagonalSizeInches = sqrt(widthInches.pow(2.0) + heightInches.pow(2.0))
-      return if (diagonalSizeInches in 3.0..6.9) {
+      val bounds: Rect = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+        windowManager.maximumWindowMetrics.bounds
+      } else {
+        val metrics = DisplayMetrics()
+        windowManager.defaultDisplay.getRealMetrics(metrics)
+
+        Rect(0, 0, metrics.widthPixels, metrics.heightPixels)
+      }
+
+      val smallestWidth: Float = dpiFromPx(
+        bounds.width().coerceAtMost(bounds.height()),
+        context.resources.configuration.densityDpi
+      )
+
+      return if (smallestWidth < MIN_TABLET_SIZE_DP) {
         // Devices in a sane range for phones are considered to be phones.
         DeviceType.PHONE
-      } else if (diagonalSizeInches > 6.9 && diagonalSizeInches <= 18.0) {
-        // Devices larger than a phone and in a sane range for tablets are tablets.
-        DeviceType.TABLET
       } else {
-        // Otherwise, we don't know what device type we're on.
-        DeviceType.UNKNOWN
+        // Devices larger than MIN_TABLET_SIZE_DP are considered tablets
+        DeviceType.TABLET
       }
     }
+
+    private fun dpiFromPx(size: Int, densityDpi: Int): Float {
+      val densityRatio = densityDpi.toFloat() / DisplayMetrics.DENSITY_DEFAULT
+      return size / densityRatio
+    }
```

Comment:
Android docs suggest never to implement this equation and to use https://developer.android.com/reference/android/util/TypedValue#applyDimension(int,%20float,%20android.util.DisplayMetrics) instead (PX to DIP in this case).